_posts/2025-05-16-last-ten-years.markdown

```diff
@@ -0,0 +1,137 @@
+In particular, we'd already made two of the biggest changes from the first version of CHERI that I had used.
+
+In the very first CHERI version, capability bounds were represented by a base and a top, but there was no separate address.
+In C, pointer arithmetic would move the base (truncating the capability) so you often needed to carry a separate offset.
+Around 2014, we changed this so the capability stored an offset instead.
+Microarchitecturally, this was actually an address, but we exposed it in the ISA as an offset from the base.
+
+Exposing this address as an offset was my idea.
+I hoped that we would be able to create a C dialect where copying garbage collectors were possible.
+Any conversion from a pointer to an integer gave the offset unless you intentionally asked for the address (in which case it was your responsibility to ensure that you didn't rely on that address being stable across GC sweeps).
+This meant that (in theory) you could relocate any object and update all capabilities to it (and, remember, the tag bit means that you *can* find all capabilities).
+I had a student, Munraj Vadera, write a copying GC for C and it worked!
+It could even do sub-object collection, so if you allocated an array and kept a bounded capability to a single element, the rest of the array could be collected.
+
+I was very happy with this model.
+Unfortunately, it turned out to be a terrible idea.
+When Alex Richardson and Jessica Clarke tried to build much more complex programs with it, they discovered that lots of things really want to use pointer to integer conversions for things like trees and hash tables.
+There were some really nasty corner cases, such as hash tables that 'worked', but ended up putting every object in the first hash bucket, giving truly awful performance.
+
+Alex and Jess fixed that in the compiler, but the core hardware abstractions remained largely unchanged.
+In CHERIoT, we removed the last vestiges of offsets by removing the set-offset instruction, because the extra addition hurt critical path lengths.
+I still believe that this was the right thing to try for research, but it's not research if you know it will work in advance, and this one didn't work at all.
+
+The other big change that was in before the paper was to generalise the sealing mechanism.
+The [first public version of the CHERI ISA](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-850.pdf) contained two instructions for sealing, `CSealCode` and `CSealData` (pages 97 and 99).
+These were intended to create cross-domain entry points.
+A cross-compartment call would be represented by two capabilities sealed with the same type.
+These would be installed in the PCC and an invoked-data capability register during a compartment call.
+
+By the 2015 paper, these were replaced by a single `CSeal` instruction that *could* seal pairs of capabilities in this way but could also seal capabilities independently.
```

Comment:
It could be worth citing Alex Richardson’s PhD dissertation for the sentry work, and possibly also for the observation on offsets vs addresses, since that’s one of the evaluation case studies in his PhD.